lib/gds_api/locations_api.rb

```diff
@@ -0,0 +1,30 @@
+require_relative "base"
+require_relative "exceptions"
+
+class GdsApi::LocationsApi < GdsApi::Base
+  # Get a list of local custodian codes for a postcode
+  #
+  # @param [String, nil] postcode The postcode for which the custodian codes are requested
+  #
+  # @return [Array] All local custodian codes for a specific postcode
+  def local_custodian_code_for_postcode(postcode)
+    response = get_json("#{endpoint}/locations?postcode=#{postcode}.json")
+
+    return [] if response.nil?
+
+    response.map { |r| r["local_custodian_code"] }
+  end
+
+  # Get a list of coordinates for a postcode
+  #
+  # @param [String, nil] postcode The postcode for which the coordinates are requested
+  #
+  # @return [Array] A list o coordinates (two fields, "latitude" and "longitude") for a specific postcode
+  def coordinates_for_postcode(postcode)
+    response = get_json("#{endpoint}/locations?postcode=#{postcode}.json")
+
+    return [] if response.nil?
+
+    response.map { |r| { "latitude" => r["latitude"], "longitude" => r["longitude"] } }
```

Comment:
@brucebolt just thinking back to our handover discussion.  It makes sense for `local_custodian_code_for_postcode` to return multiple custodian codes, as a user's postcode could be split across multiple authorities, so we want to surface that.

But for `coordinates_for_postcode`, am I right in remembering that we wanted to return one set of coordinates, i.e. return the first result in the list, or do some work to approximate the average coordinates across all the results in the response? I'm afraid we didn't specify in the [Trello card](https://trello.com/c/wufIa6zM/2847-add-locations-api-to-gds-api-adapters-3), so just double checking 😅

Reply:
As we didn't want to POST back the user's exact address when a user selects their full address (otherwise we get into PII collection issues), we'll have to use a single set of coordinates for all addresses in the postcode.  For postcodes where there is a single local authority, we'll have to do this anyway.

I don't believe the exact location is too important.  As far as I can see, we don't report the actual distance to users when doing imminence lookups, so an approximation should be suitable to order the list by nearest.

As for choosing the first or taking an average, I think an average (if we can do this) would be better as this should give us roughly the centre of the postcode (assuming the buildings are all evenly distributed).

Reply:
Thanks @brucebolt for the context. I'll change the method to return the average.

Reply:
Thanks for clarifying, @brucebolt. I think what we return so far follows the [architecture plan (see step 5)](https://docs.google.com/document/d/1p29g1SQgi2obQnPPsl9amx7xmrJonyH4tRbVMAhXy-A/edit#heading=h.7v5d22d02otc), but it sounds like we'll need a small change.

We probably ought to do the averaging in Locations API (rather than GDS API Adapters), and could return that at the top level of the response, i.e. 1):

```diff
- [ { lat: 1, long: -2, address: "Foo" }, { lat: 2, long: -3, address: "Bar" } ]
+ { lat: 1.5, long: -2.5, [ { lat: 1, long: -2, address: "Foo" }, { lat: 2, long: -3, address: "Bar" } ] }
```

But I think you're saying we need to strip out the coordinates from each address? So it would be more like 2):

```diff
- [ { address: "Foo" }, { address: "Bar" } ]
+ { lat: 1.5, long: -2.5, [ { address: "Foo" }, { address: "Bar" } ] }
```

And if we're doing that, we could just repeat the top-level averages inside each result and keep the same array response structure 3):

```diff
-  { lat: 1.5, long: -2.5, [ { address: "Foo" }, { address: "Bar" } ] }
+ [ { lat: 1.5, long: -2.5, address: "Foo" }, { lat: 1.5, long: -2.5, address: "Bar" } ]
```

Can I check we're on the same page and you're suggesting option 3?

Reply:
I wonder if we should go with option 1?

This would allow us to increase the scope of Locations API if needed in the future (e.g. we build a new service that requires coordinates for each property) but solves the immediate problem of needing to locate (roughly) the centre of a postcode.

Reply:
Sounds good to me, so long as there are no PII implications for keeping the coordinates in (I wouldn't have thought there's any increased risk, as we're already returning the full address in the Locations API response).

I'll write up a quick card now 👍

Reply:
https://trello.com/c/N46nNKYN/2867-return-average-lat-long-coordinates-in-locations-api